patches/qemu-5.1.0.patch

```diff
@@ -3226,6 +3311,183 @@ index 09faa69a54..44d1f3644d 100755
+ #if defined(CONFIG_TCG_INTERPRETER)
+diff --git a/include/tcg/tcg-apple-jit.h b/include/tcg/tcg-apple-jit.h
+new file mode 100644
+index 0000000000..6ed480b634
+--- /dev/null
++++ b/include/tcg/tcg-apple-jit.h
+@@ -0,0 +1,76 @@
++/*
++ * Apple Silicon APRR functions for JIT handling
++ *
++ * Copyright (c) 2020 osy
++ *
++ * This library is free software; you can redistribute it and/or
++ * modify it under the terms of the GNU Lesser General Public
++ * License as published by the Free Software Foundation; either
++ * version 2.1 of the License, or (at your option) any later version.
++ *
++ * This library is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++ * Lesser General Public License for more details.
++ *
++ * You should have received a copy of the GNU Lesser General Public
++ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
++ */
++
++// Credits to: https://siguza.github.io/APRR/
++// Reversed from /usr/lib/system/libsystem_pthread.dylib
++
++#ifndef TCG_APPLE_JIT_H
++#define TCG_APPLE_JIT_H
++
++#if defined(__aarch64__) && defined(CONFIG_DARWIN)
++
++#define _COMM_PAGE_START_ADDRESS        (0x0000000FFFFFC000ULL) /* In TTBR0 */
++#define _COMM_PAGE_APRR_SUPPORT         (_COMM_PAGE_START_ADDRESS+0x10C)
++#define _COMM_PAGE_APPR_WRITE_ENABLE    (_COMM_PAGE_START_ADDRESS+0x110)
++#define _COMM_PAGE_APRR_WRITE_DISABLE   (_COMM_PAGE_START_ADDRESS+0x118)
++
++static __attribute__ ((__always_inline__)) bool jit_write_protect_supported(void)
++{
++    uint8_t aprr_support = *(volatile uint8_t *)_COMM_PAGE_APRR_SUPPORT;
++    return aprr_support > 0;
++}
++
++static __attribute__ ((__always_inline__)) void jit_write_protect(int enabled) // write protect enable = write disable
++{
++    uint8_t aprr_support = *(volatile uint8_t *)_COMM_PAGE_APRR_SUPPORT;
++    if (aprr_support == 0 || aprr_support > 3) {
++        return;
++    } else if (aprr_support == 1) {
++        __asm__ __volatile__ (
++            "mov x0, %0\n"
++            "ldr x0, [x0]\n"
++            "msr S3_4_c15_c2_7, x0\n"
++            "isb sy\n"
++            :: "r" (enabled ? _COMM_PAGE_APRR_WRITE_DISABLE : _COMM_PAGE_APPR_WRITE_ENABLE) : "memory", "x0"
++        );
++    } else {
++        __asm__ __volatile__ (
++            "mov x0, %0\n"
++            "ldr x0, [x0]\n"
++            "msr S3_6_c15_c1_5, x0\n"
++            "isb sy\n"
++            :: "r" (enabled ? _COMM_PAGE_APRR_WRITE_DISABLE : _COMM_PAGE_APPR_WRITE_ENABLE) : "memory", "x0"
++        );
++    }
++}
++
++#else // defined(__aarch64__) && defined(CONFIG_DARWIN)
++
++static __attribute__ ((__always_inline__)) bool jit_write_protect_supported(void)
++{
++    return false;
++}
++
++static __attribute__ ((__always_inline__)) void jit_write_protect(int enabled)
++{
++}
++
++#endif
```

Comment:
Consider using `os_thread_self_restrict_rwx_is_supported`/`os_thread_self_restrict_rwx_to_rx`, they backdeploy to earlier iOS versions.

Reply:
Where is that defined?

Reply:
Internal SDK, you'll have to forward declare it.

Reply:
Just tried and it didn't link on iOS 13.4.1

Reply:
`grep -r 'os_thread_self_restrict_rwx_to_rx' /Applications/Xcode.app` on Xcode 12 returned no results.

Reply:
Actually, it's probably a macro. Nevermind then.